### ISSUE TITLE
Difference in cypher peformance depending on phrasing?

### ISSUE THREAD

**Issue by jhb** [neo4j/neo4j#1094]
Phrase one needs 1500 - 9000ms, phrase two only 200-400ms. There shouldn't be such a big difference. Full details: http://stackoverflow.com/posts/18127799

dataset:

```
   *1                100k              *3..30
[:author]-------->  person  <---------[:member]
    |                 |                  |
    |               *2..8                |
    |              [:topic]              |
    |                 |                  |
    |    *3..8        V        *3..8     |
article-[:topic]->  topic  <-[:topic]- project
    1m                1k                 20k
```

The numbers mean e.g. I have 1 million articles, each with one author, and each article has between 3 and 8 rels to one of 1000 topics.

phrase one:

```
match 
    n:person<-[:member]-p:project-[:member]->m, 
    n-[:topic]->t,
    m-[:topic]->t 
where 
    n.name='person1215' and 
    n<>m 
return 
    m,t;
```

phrase two:

```
match 
    n:person<-[:member]-p:project-[:member]->m, 
    n-[:topic]->t1, 
    m-[:topic]->t2  
where 
    n.name='person1250' and 
    n<>m and 
    t1=t2 
return 
    t1,m;
```

**Comment by sherfert** [comment]
Thanks for reporting that. The planner introduces an unnecessary Cartesian product for query 1. We will investigate this.

**Comment by sherfert** [comment]
This has been solved in Neo4j 3.4.
